### DISCUSSION TITLE
Continuation of issue #2 (OpenArena, etc.)

### DISCUSSION THREAD

**Discussion by torhus** [torhus/monster-browser#22] [Ideas]
> First thing I though that could be added is an ".exe localizer" (or better said, a path router) so you can point to monsterbrowser where the .exe for each game/mod are. It also could ask you if you try to launch a game and can't find the .exe ("Monsterbrowser was not able to locate whatever.exe, want to locate .exe? Yes/No" smth).

I think at least a better error message would help, I will see about doing more.

> EDIT: Sorry for the edit. Could be possible that monsterbrowser read again mods.ini before launching the game? That way we could avoid to restart it everytime we do changes on mods.ini (it's not likely you are going to do changes all the time but i think it could help).

It does load mods.ini again if you go through the settings dialog when you edit, at least. It stores the file's modification time when you click the Game configuration button, then checks if it changed when you close the dialog.

**Comment by ghost** [comment]
> It does load mods.ini again if you go through the settings dialog when you edit, at least. It stores the file's modification time when you click the Game configuration button, then checks if it changed when you close the dialog.

I think I understand it more or less but when I do "Load game - Error - Close Error - modify .ini - Load game again" it should run the modified .ini but it gives me the same error (in this case, .exe was not locate).

EDIT: Ok I tried your way and it works, it just that I was doing it opening .ini "manually". I still think that have both options could be nice but it's a minor thing :D

**Reply by The-Gig** [reply]
Of course it detects changes to the .ini file if you pass through the program settings window. Detecting changes done editing the .ini file opening it from Explorer would require either a cycle checking for it every x time or checking the file every time one presses any of the controls in the upper bar...

**Reply by ghost** [reply]
Couldn't be possible to re-check after you click on launch a game and before it tries to launch it?

**Reply by torhus** [reply]
There are many cases where this would be nice to have, the question is where to stop...

**Reply by torhus** [reply]
Automatic reloading of mods.ini was easy to do, so it's coming in the next release. It's in this experimental version here: https://ci.appveyor.com/project/torhus/monster-browser/builds/45110848/artifacts

**Reply by The-Gig** [reply]
I just tested it: I opened the .ini file in Notepad directly from Windows Explorer, then started MB, then made changes to _section name_ and/or _mod_ value in the ini file and saved. MB automatically updated the results.

But sometimes I experienced some crashes, with _MB closing down by itself_. I'm not 100% sure about how to reproduce reliably, but it looks like it mostly happens in case you change (in the ini file) the section name of the game currently selected in MB, then save, _then save again, even without making further changes_.

**Reply by torhus** [reply]
Thanks for testing, fixed the crash!

**Reply by The-Gig** [reply]
You're welcome! :-)

**Comment by The-Gig** [comment]
Dropping a link to the "issue" this "discussion" is a continuation to: https://github.com/torhus/monster-browser/issues/2

Also, for further continuing the bit about "quick filter" (at the moment, typing a word in the game/mod field which does not correspond to any of the entries of the dropdown list) working only with the main master server (see these posts: https://github.com/torhus/monster-browser/issues/2#issuecomment-1236820051, https://github.com/torhus/monster-browser/issues/2#issuecomment-1236996531)... should we talk about it here, or open an apposite Discussion?

**Reply by torhus** [reply]
At least there are threads here. I think the current quick filter makes little sense, maybe it should be removed? Would some kind of search field that would just show all servers that have a name, cvar value, or player that matches the search make sense?

**Reply by The-Gig** [reply]
I'm not really sure about how threads work here (first time using github "Discussions" feature), however I think some kind of "search" feature in MB would be useful, I even proposed a similar idea some time ago for players and possibly more: https://github.com/torhus/monster-browser/issues/5 (I know, ideas are "cheap" and actually doing things is the matter...)

**Reply by ghost** [reply]
> I'm not really sure about how threads work here (first time using github "Discussions" feature), however I think some kind of "search" feature would be useful, I even proposed a similar idea some time ago for players and possibly more: #5 (I know, ideas are "cheap" and actually doing things is the matter...)

It's more a discussion, like the name implies. Issues are for report new stuff or report bugs without the "discussion" part :D.

**Reply by The-Gig** [reply]
Searching into _all_ cvar values is something I didn't think about... it might be useful, but maybe the game should not search in there by default? I'm not sure.
So, maybe a text field accompained by a dropdown list which allows you to choose where to search into?
A list like "player name, server name, [server address?], map, gametype (type number or description in the text field), [game?], [gamename?], game+gamename, [custom cvar (type cvarname:string in the text field) ?], all cvars, [everywhere ?], [everywhere except cvars ?]"?

In the list above I've put everything I could think of, it might be shorter. I've put square brackets around the most expendable ones.

**Reply by ghost** [reply]
> Searching into _all_ cvar values is something I didn't think about... it might be useful, but maybe the game should not search in there by default? I'm not sure. So, maybe a text field accompained by a dropdown list which allows you to choose where to search into? A list like "player name, server name, [server address?], map, gametype (type number or description in the text field), [game?], [gamename?], game+gamename, [custom cvar (type cvarname:string in the text field) ?], all cvars, [everywhere ?], [everywhere except cvars ?]"?
> 
> In the list above I've put everything I could think of, it might be shorter. I've put square brackets around the most expendable ones.

sounds like an awesome idea. I remember there was an urban terror dedicated browser that had something like this but I can't recall the name of that software, maybe I can try to locate it and take some ideas from it

**Reply by torhus** [reply]
I added a basic version of this feature, just a field that lets you filter on server name and game/gamename cvars:
https://ci.appveyor.com/project/torhus/monster-browser/builds/44720025

Seems useful for finding the servers you are interested in, what do you think?

**Reply by The-Gig** [reply]
Tried (beta Sep 8 2022). Very good! On a side note, maybe some kind of indication about what that field is for may be useful for newcomers.

**Comment by torhus** [comment]
I did some work on the UI for this, does this look ok? Player search is the next thing to try.

[MonsterBrowser_filtertest.zip](https://github.com/torhus/monster-browser/files/9749285/MonsterBrowser_filtertest.zip)

![MB_filtertest](https://user-images.githubusercontent.com/19881605/194953998-57e4d8e5-dc17-43a7-b6c9-1f70db63b9df.png)

**Reply by The-Gig** [reply]
Here I am!
So, against the previus test I had (Sep 8 2022), this test (Oct 10 2022) just moves to the left the game selection list? Wonder why it's got from 2.9 MB to 4.8 MB...

Anyway...
I tested the search feature for server name and game and gamename and it seems to work. :-)
What I would suggest about it is:
- To make the search field _appear longer_ (IDK, maybe double the lenght?). At the moment, even typing there a simple "quake3arena" cuts part of the "q".
- To make the search field recognize what you typed _as soon as you click outside from it_. At the moment you _need_ to press Enter key to have the program register the input... if you forget about it, you end up with the field showing a certain text, while the filter actually applied is still the previous one! Which could be confusing or annoying...

**Reply by torhus** [reply]
> Here I am! So, against the previus test I had (Sep 8 2022), this test (Oct 10 2022) just moves to the left the game selection list? Wonder why it's got from 2.9 MB to 4.8 MB...

It's a 64-bit build, the actual release will probably still be 32 bits, should be a bit smaller. Different compiler version and libraries, too.

> 
> Anyway... I tested the search feature for server name and game and gamename and it seems to work. :-) What I would suggest about it is:
> 
>     * To make the search field _appear longer_ (IDK, maybe double the lenght?). At the moment, even typing there a simple "quake3arena" cuts part of the "q".
> 

Sure, it's a bit short now.

>     * To make the search field recognize what you typed _as soon as you click outside from it_. At the moment you _need_ to press Enter key to have the program register the input... if you forget about it, you end up with the field showing a certain text, while the filter actually applied is still the previous one! Which could be confusing or annoying...

I'm considering just having it update automatically every second or so.

**Reply by torhus** [reply]
Updating the search automatically now, just the players search left...
https://ci.appveyor.com/project/torhus/monster-browser/builds/45039184/artifacts

**Reply by The-Gig** [reply]
Tried. Seems to work well!
Althought maybe I would have made the field still a bit longer.
IDK if you planned to also search into the IP field... at the moment it doesn't do that.

Side notes:
- I noticed that when you check "Not empty", it automatically unchecks "Have humans". Is it on purpose?
- Now that the filter group takes two lines of space, what about adding a "Not full" checkbox? Various game browsers have a such feature.
- Also, wondering if now you might have the room to write "Check for new servers" instead of just "Check for new" in the apposite button, as at the moment someone might think it stands for "Check for new MonsterBrowser version"...

**Reply by The-Gig** [reply]
Just a reminder that in 0.9h the "problem" about that checkbox which gets unchecked (see "side notes" in the post above) is still there.

**Reply by torhus** [reply]
Yes, the check box is unchecked on purpose, so you don't have to do it yourself.

**Comment by The-Gig** [comment]
About OpenArena, _maybe_ there are currently enough servers running either "aftershock" or "rat" mods to be worth being added as pre-set "games" in the dropdown menu.

As per their _gametypes descriptions_, those for "aftershock" seem to be the same as baseoa. The latest official release of Aftershock XE is 229, I just tried it and Possession has not been backported to it yet (gametype 13 falls back to FFA). <https://github.com/Irbyz/aftershock-xe>

For "rat" mod, gametypes up to 12 _seem_ to be the same as baseoa, while gametype 13 seems to be "treasure hunter". I connected to a server running gametype 14 and the loading screen mentioned "multitournament" as gametype, while later trying the gametype 14 locally said "unknown gametype" and showed a team-based thing (instead of a non-team-based thing as it happened online), so there's something strange (probably 14 requires some extra configuration). And their help page about gametypes (<https://ratmod.github.io/game-types.html>) also mentions an "Xterm" mode which I have not been able to find (maybe it's just a modifier for gametype 8). But the main page in their documentation page (<https://ratmod.github.io/index.html>) mentions "new gametypes: Multitournament (multiple simultaneous duels), Extermination (Wipeout), FreezeTag, CoinFFA, CoinTDM, Treasure Hunter". In case you plan to add a preset for ratmod, I can try asking for infos on their Discord.

**Reply by torhus** [reply]
Right. I'm hoping that people in the game's community will create a good configuration and post it on their forums, that's partly why it's a text file.

**Reply by The-Gig** [reply]
The "unknown gametype" thing was due to me having an old pk3 file which conflicted with the current one. Anyway, even after fixing that issue, there is still one thing which puzzles me about rat's gametypes 13 and 14, I'm waiting for its community for some clarifications. (Also, they said most of the other mentioned game modes are _mutators_ for existing gametypes.)

In the meanwhile, can I ask you if you can see the "/N/" servers (which run "rat" mod) in MonsterBrowser? Yesterday I was seeing them in the list, then at a certain point they disappeared. I thought maybe they were having some tech problems and waited for today. Today, I still can't see them in Monster Browser, so I tried with Qtracker and they are not listed even there. But they are listed in the dpmaster website <http://dpmaster.deathmask.net/?game=openarena> and in the OA built-in server browser, and I can connect to them. So I'd like to find out if the anomaly is only mine or also affects others...

**Reply by The-Gig** [reply]
So, rat gametype 13 is actually "Treasure Hunter", it's just a small bug which erroneously shows "multitournament" in the map loading screen. Multitournament is gametype 14.
So, if you like, I think you can create presets for AfterShock (same gametypes are baseoa, up to 12 -Domination-) and Ratmod (same gametypes as baseoa up to 12, then "TH", then "MT" _I guess_?

I still have no clue about the reason why /N/ servers do not reply to my queries using qstat/oaquery/qtracker/monsterbrowser, while I can actually connect to them. Do you see them, using such tools?

**Reply by The-Gig** [reply]
About those servers, I don't have an official explanation, but I heard that some OA servers got DOS attacks lately, so it's probable the admins of those servers added some kind of defensive mechanism to block third party queries. I hope that's just temporary... I think you can go on serenely.

**Reply by torhus** [reply]
Ok, thanks. I'm not sure I will add mods for OpenArena, trying not to clutter the list too much :)
